client.go

```diff
@@ -599,6 +604,67 @@ func (c *client) SubscribeMultiple(filters map[string]byte, callback MessageHand
 	return token
 }
 
+// Load all stored messages and resend them
+// Call this to ensure QOS > 1,2 even after an application crash
+func (c *client) resume(subscription bool) []Token {
+
+	storedKeys := c.persist.All()
+	tokens := make([]Token, len(storedKeys))
+	for idx, key := range storedKeys {
+		packet := c.persist.Get(key)
+		details := packet.Details()
+		if isKeyOutbound(key) {
+			switch packet.(type) {
+			case *packets.SubscribePacket:
+				if subscription {
+					DEBUG.Println(STR, fmt.Sprintf("loaded pending subscribe (%d)", details.MessageID))
+					token := newToken(packets.Subscribe).(*SubscribeToken)
+					tokens[idx] = token
```

Comment:
I'm unclear why the list of tokens returned is not a slice with no nils?  That is to say, why not simply append tokens as needed?  The code as written seems to produce some Tokens which are nil.

Reply:
Indeed this array may contains nil token. At the time I choose this to avoid memory allocation behind the append function (if I understood correctly how the append function works). I think it costs more allocating memory than checking if an element is nil. But I agree it will be more convenient if there was no nil. I can change it if convenience is prefered to performance.

Reply:
It should be an extremely rare operation, with a lot of cost (such as pulling things off of disk) anyway.  So I would prioritize a clean interface over saving a few cycles.

Reply:
In fact in my use case it is not rare. I am using this on a mobile device which is not always connected. This feature allows me to backlog all published message on the go once returned near the station where it reconnects to the broker. As my device is battery powered I tend to minimize usage of cpu. But I guess my usage is quiet specific and not suited for a generic library. I will make the changes depending of the result of the discussion about what to do with this array.